stdlib/public/runtime/LibPrespecialized.cpp

```diff
@@ -45,6 +44,8 @@ static bool prespecializedLoggingEnabled = false;
       fprintf(stderr, "Prespecializations library: " fmt "\n", __VA_ARGS__);   \
   } while (0)
 
+#define LOG0(string) LOG("%s", string)
```

Comment:
Actually, I _do_ see a way. How about

```
#define LOG(fmt, ...)                                                         \
  do {                                                                        \
    if (SWIFT_UNLIKELY(prespecializedLoggingEnabled))                         \
      log("Prespecializations library: " fmt "\n" __VA_OPT__(,) __VA_ARGS__); \
  } while (0);

static inline void log(const char *msg) {
  fputs(msg, stderr);
}

template <typename ...Ts>
static inline void log(const char *fmt, Ts... args) {
  fprintf(stderr, fmt, args...);
}
```

Reply:
Of course, that does mean using `__VA_OPT__`, which I _think_ is OK for Clang and GCC even before C++20. Or we could use the `##` extension.

Reply:
I'd be fine with requiring the zero-argument log call to not use format specifiers, so we could skip the overloaded `log`. But I'm wary of `__VA_OPT__`. I think we need to build with MSVC sometimes and support there seem to be iffy from my searching.

We could write it as just a static inline function with a variadic template and no macro at all.

Reply:
You can't get the compiler to add the trailing `\n` to the string literal without a macro so that's a little annoying.

An alternate approach: function parameter lists don't accept trailing commas, but initializer lists do, so write the log function as a constructor on a helper type:

```
template <typename ...Args>
struct Log {
  Log(const char *fmt, Args... args) {
    fprintf(stderr, fmt, args...);
  }
};

#define LOG(fmt, ...) Log{fmt "\n", __VA_ARGS__}
```

Reply:
The macro wrapper is useful, I think, because it lets you avoid evaluating the arguments when logging is disabled. I like your thinking about the initializer list; that should work nicely.

Reply:
You're right, we do need a macro regardless.

The initializer list ends up with a lot of ceremony to avoid warnings and I don't think it's worth it just to avoid a `0` variant. It would be worthwhile if we want to introduce a standardized logging macro for the whole runtime at some point. For posterity, this is what I ended up with:

```
struct LogHelper {
#ifdef __clang__
  __attribute((__format__(__printf__, 2, 3)))
#endif
  LogHelper(const char *fmt, ...) {
    va_list args;
    va_start(args, fmt);
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wformat-security"
#pragma clang diagnostic ignored "-Wformat-nonliteral"
    vfprintf(stderr, fmt, args);
#pragma clang diagnostic pop
    va_end(args);
  }
};

#pragma clang diagnostic ignored "-Wgnu-zero-variadic-macro-arguments"

#define LOG(fmt, ...)                                                          \
  do {                                                                         \
    if (SWIFT_UNLIKELY(prespecializedLoggingEnabled)) {                        \
      LogHelper{"Prespecializations library: " fmt "\n", __VA_ARGS__};   \
    } \
  } while (0)
```

Reply:
It was an interesting experiment, at least. :-)